File: clients/quarkus/flowcatalyst-sdk/src/main/java/tech/flowcatalyst/sdk/client/resources/EventTypes.java
package tech.flowcatalyst.sdk.client.resources;

import com.fasterxml.jackson.core.type.TypeReference;
import tech.flowcatalyst.sdk.client.FlowCatalystClient;
import tech.flowcatalyst.sdk.dto.EventType;
import tech.flowcatalyst.sdk.dto.ListResult;
import tech.flowcatalyst.sdk.enums.SchemaType;

import java.util.List;
import java.util.Map;

/**
 * Resource for managing event types.
 */
public class EventTypes {

    private final FlowCatalystClient client;

    public EventTypes(FlowCatalystClient client) {
        this.client = client;
    }

    /**
     * List all event types with optional filters.
     */
    public ListResult<EventType> list() {
        return list(Map.of());
    }

    /**
     * List event types with filters.
     *
     * @param filters Optional filters: status, application, subdomain, aggregate
     */
    public ListResult<EventType> list(Map<String, String> filters) {
        String query = filters.isEmpty() ? "" : "?" + buildQuery(filters);
        var response = client.request("GET", "/api/event-types" + query, null,
            new TypeReference<Map<String, Object>>() {});

        @SuppressWarnings("unchecked")
        var items = (List<Map<String, Object>>) response.getOrDefault("items", List.of());
        var eventTypes = items.stream()
            .map(this::mapEventType)
            .toList();

        return ListResult.of(eventTypes);
    }

    /**
     * Get an event type by ID.
     */
    public EventType get(String id) {
        var response = client.request("GET", "/api/event-types/" + id, null,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Create a new event type.
     */
    public EventType create(CreateEventTypeRequest request) {
        var response = client.request("POST", "/api/event-types", request,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Update an event type.
     */
    public EventType update(String id, UpdateEventTypeRequest request) {
        var response = client.request("PATCH", "/api/event-types/" + id, request,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Add a schema version to an event type.
     */
    public EventType addSchema(String id, AddSchemaRequest request) {
        var response = client.request("POST", "/api/event-types/" + id + "/schemas", request,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Finalise a schema version (FINALISING -> CURRENT).
     */
    public EventType finaliseSchema(String id, String version) {
        var response = client.request("POST",
            "/api/event-types/" + id + "/schemas/" + version + "/finalise", null,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Deprecate a schema version (CURRENT -> DEPRECATED).
     */
    public EventType deprecateSchema(String id, String version) {
        var response = client.request("POST",
            "/api/event-types/" + id + "/schemas/" + version + "/deprecate", null,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Archive an event type.
     */
    public EventType archive(String id) {
        var response = client.request("POST", "/api/event-types/" + id + "/archive", null,
            new TypeReference<Map<String, Object>>() {});
        return mapEventType(response);
    }

    /**
     * Delete an event type.
     */
    public void delete(String id) {
        client.requestVoid("DELETE", "/api/event-types/" + id, null);
    }

    /**
     * Get distinct application names for filtering.
     */
    public List<String> filterApplications() {
        var response = client.request("GET", "/api/event-types/filters/applications", null,
            new TypeReference<Map<String, Object>>() {});

        @SuppressWarnings("unchecked")
        var apps = (List<String>) response.getOrDefault("applications", List.of());
        return apps;
    }

    /**
     * Get distinct subdomains for filtering.
     */
    public List<String> filterSubdomains(String application) {
        String query = application != null ? "?application=" + application : "";
        var response = client.request("GET", "/api/event-types/filters/subdomains" + query, null,
            new TypeReference<Map<String, Object>>() {});

        @SuppressWarnings("unchecked")
        var subdomains = (List<String>) response.getOrDefault("subdomains", List.of());
        return subdomains;
    }

    private EventType mapEventType(Map<String, Object> data) {
        return client.getObjectMapper().convertValue(data, EventType.class);
    }

    private String buildQuery(Map<String, String> params) {
        return params.entrySet().stream()
            .map(e -> e.getKey() + "=" + e.getValue())
            .reduce((a, b) -> a + "&" + b)
            .orElse("");
    }

    // Request DTOs

    public record CreateEventTypeRequest(
        String code,
        String name,
        String description
    ) {}

    public record UpdateEventTypeRequest(
        String name,
        String description
    ) {}

    public record AddSchemaRequest(
        String version,
        String mimeType,
        String schema,
        SchemaType schemaType
    ) {}
}

File: packages/platform-ui-vue/src/api/event-types.ts
import { bffFetch } from './client';

// Types (will be replaced by Hey API generated types)
export type EventTypeStatus = 'CURRENT' | 'ARCHIVE';
export type SchemaType = 'JSON_SCHEMA' | 'PROTO' | 'XSD';
export type SpecVersionStatus = 'FINALISING' | 'CURRENT' | 'DEPRECATED';

export interface SpecVersion {
  version: string;
  mimeType: string;
  schema?: string;
  schemaType: SchemaType;
  status: SpecVersionStatus;
}

export interface EventType {
  id: string;
  code: string;
  application: string;
  subdomain: string;
  aggregate: string;
  event: string;
  name: string;
  description?: string;
  status: EventTypeStatus;
  clientScoped: boolean;
  specVersions: SpecVersion[];
  createdAt: string;
  updatedAt: string;
}

export interface EventTypeListResponse {
  items: EventType[];
  total: number;
}

export interface FilterOptionsResponse {
  options: string[];
}

export interface CreateEventTypeRequest {
  code: string;
  name: string;
  description?: string;
  clientScoped: boolean;
}

export interface UpdateEventTypeRequest {
  name?: string;
  description?: string;
}

export interface AddSchemaRequest {
  version: string;
  mimeType: string;
  schema: string;
  schemaType: SchemaType;
}

export interface EventTypeFilters {
  applications?: string[];
  subdomains?: string[];
  aggregates?: string[];
  status?: EventTypeStatus;
}

// API functions - using BFF endpoints for JavaScript-safe string IDs
export const eventTypesApi = {
  list(filters: EventTypeFilters = {}): Promise<EventTypeListResponse> {
    const params = new URLSearchParams();
    filters.applications?.forEach(a => params.append('application', a));
    filters.subdomains?.forEach(s => params.append('subdomain', s));
    filters.aggregates?.forEach(a => params.append('aggregate', a));
    if (filters.status) params.set('status', filters.status);

    const query = params.toString();
    return bffFetch(`/event-types${query ? `?${query}` : ''}`);
  },

  get(id: string): Promise<EventType> {
    return bffFetch(`/event-types/${id}`);
  },

  create(data: CreateEventTypeRequest): Promise<EventType> {
    return bffFetch('/event-types', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  update(id: string, data: UpdateEventTypeRequest): Promise<EventType> {
    return bffFetch(`/event-types/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    });
  },

  delete(id: string): Promise<void> {
    return bffFetch(`/event-types/${id}`, { method: 'DELETE' });
  },

  archive(id: string): Promise<EventType> {
    return bffFetch(`/event-types/${id}/archive`, { method: 'POST' });
  },

  addSchema(id: string, data: AddSchemaRequest): Promise<EventType> {
    return bffFetch(`/event-types/${id}/schemas`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  finaliseSchema(id: string, version: string): Promise<EventType> {
    return bffFetch(`/event-types/${id}/schemas/${version}/finalise`, {
      method: 'POST',
    });
  },

  deprecateSchema(id: string, version: string): Promise<EventType> {
    return bffFetch(`/event-types/${id}/schemas/${version}/deprecate`, {
      method: 'POST',
    });
  },

  // Filter options
  getApplications(): Promise<FilterOptionsResponse> {
    return bffFetch('/event-types/filters/applications');
  },

  getSubdomains(applications?: string[]): Promise<FilterOptionsResponse> {
    const params = new URLSearchParams();
    applications?.forEach(a => params.append('application', a));
    const query = params.toString();
    return bffFetch(`/event-types/filters/subdomains${query ? `?${query}` : ''}`);
  },

  getAggregates(applications?: string[], subdomains?: string[]): Promise<FilterOptionsResponse> {
    const params = new URLSearchParams();
    applications?.forEach(a => params.append('application', a));
    subdomains?.forEach(s => params.append('subdomain', s));
    const query = params.toString();
    return bffFetch(`/event-types/filters/aggregates${query ? `?${query}` : ''}`);
  },
};
